src/cachemachine/automatedteller.py

```diff
@@ -0,0 +1,75 @@
+import asyncio
+from typing import Any, Dict, List, Sequence, Union
+
+import structlog
+
+from cachemachine.cachechecker import CacheChecker
+from cachemachine.cachedepositer import CacheDepositer
+from cachemachine.rubinrepoman import RubinRepoMan
+from cachemachine.simplerepoman import SimpleRepoMan
+
+logger = structlog.get_logger(__name__)
+
+
+class AutomatedTeller:
+    available_images: List[Dict[str, str]] = []
+    desired_images: List[Dict[str, str]] = []
+    images_to_cache: List[Dict[str, str]] = []
+
+    def __init__(
+        self,
+        name: str,
+        labels: Dict[str, str],
+        repomen: Sequence[Union[SimpleRepoMan, RubinRepoMan]],
+    ):
+        self.name = name
+        self.labels = labels
+        self.repomen = repomen
+
+        self.recommended_image_urls = []
+        for r in self.repomen:
+            if r.recommended_image_url:
+                self.recommended_image_urls.append(r.recommended_image_url)
+
+        self.checker = CacheChecker(self.labels, self.recommended_image_urls)
+        self.depositer = CacheDepositer(self.name, self.labels)
+
+    # Note, doesn't actually return, intended to run forever.
+    async def do_work(self) -> None:
+        while True:
+            self.available_images = []
+            self.desired_images = []
+            self.images_to_cache = []
```

Comment:
Oh, I see, you want to clear these on each cycle through the loop.

I see that they're defined as instance variables because they're also used in `talk`. But I think there's something a little weird here in that `talk` is going to get either a calculated answer or the empty array or a partial result depending on when it's called, because it returns the same variables as are being actively changed by `do_work`. That seems likely to cause problems.

I'm not sure what calls `talk` and when it's used, but given that `do_work` is running asynchronously and simultaneously with `talk`, I think it needs to use variables local to that method to gather its work and then update the instance variables only when it's done, inside a lock, and `talk` also needs to use a lock.

Reply:
I thought that with async it's all one thread and you don't need to lock everything as long as you aren't doing an await inside of it?

Talk is called from the web parts, but I think those are handled on the same thread?

Reply:
I was afraid that aiojobs would use a separate thread pool for background tasks, but it looks like it doesn't. (Although note that means the entire cachemachine application is going to block on all Kubernetes operations, since those are done synchronously. Probably doesn't matter for this purpose.)

So yeah, it probably doesn't matter since there isn't an internal synchronization point, although I would tend to use local variables and then copy them into the instance variables in one place just out of paranoia in case there's some reason to add a synchronization point later (such as an async Kubernetes library). But you're right that regardless there shouldn't be a need for a lock unless we start spawning separate thread pools, and there's no reason to do that.

Reply:
Right, and I personally like to avoid unnecessary locking.  It makes everyone confused.  I can use a local and copy it over, I had it that way at one point, but then thought about it and went this way.